chapter4/src/4_2_5/main.rs

```diff
@@ -0,0 +1,314 @@
+///
+///     tokio::spawn(async move {
+///         println!("sub() is finished");
+///         canceler.cancel();
+///     });
+///
+///     let done = ctx.done().await;
+///     assert_eq!(done.unwrap_err(), ContextError::Canceled);
+///     println!("all tasks are finished");
+/// }
+/// ```
+struct ContextWithCancel {
+    cancel_notify: Notify,
+    tree_node: Arc<Mutex<ContextTree>>,
+}
+
+impl ContextWithCancel {
+    pub fn new<C: 'static + HasContextTree + Context>(
+        context: Arc<C>,
+    ) -> (Arc<Self>, Canceler<Self>) {
+        let c: Arc<dyn Context> = context.clone();
+        let this = Arc::new(Self {
+            cancel_notify: Notify::new(),
+            tree_node: Arc::new(Mutex::new(ContextTree {
+                canceled: None,
+                children: vec![],
+                parent: Some(Arc::downgrade(&c)),
+            })),
+        });
+        context
+            .context_tree()
+            .lock()
+            .unwrap()
+            .children
+            .push(this.clone());
+        (this.clone(), Canceler { context: this })
+    }
+
+    pub async fn done(&self) -> Result<(), ContextError> {
+        let _ = self.cancel_notify.notified().await;
+        if let Some(e) = self.tree_node.lock().unwrap().canceled.clone() {
+            Err(e)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Contextのcancelを実際に行うための構造体
+/// # Example
+///
+/// ```
+///
+/// #[tokio::main]
+/// async fn main() {
+///     println!("start sub()");
+///
+///     let (ctx, canceler: Canceler) = ContextWithCancel::new(BackgroundContext::new()); // Contextと一緒にCancelerが返ってくる
+///
+///     tokio::spawn(async move {
+///         println!("sub() is finished");
+///         canceler.cancel();
+///     });
+///
+///     let done = ctx.done().await;
+///     assert_eq!(done.unwrap_err(), ContextError::Canceled);
+///     println!("all tasks are finished");
+/// }
+/// ```
+struct Canceler<C: CancelPropagate> {
+    context: Arc<C>,
+}
+
+impl<C: CancelPropagate> Canceler<C> {
+    /// Contextのcancelを実際に行う関数
+    pub fn cancel(&self) {
+        self.context.cancel_propagate(ContextError::Canceled)
+    }
+}
+
+impl HasContextTree for ContextWithCancel {
+    fn context_tree(&self) -> Arc<Mutex<ContextTree>> {
+        self.tree_node.clone()
+    }
+}
+
+impl Context for ContextWithCancel {
+    fn done(&self) -> Pin<Box<dyn Future<Output = Result<(), ContextError>> + '_>> {
+        Box::pin(ContextWithCancel::done(self))
+    }
+
+    fn err(&self) -> Option<ContextError> {
+        self.tree_node.lock().unwrap().canceled.clone()
+    }
+
+    fn value(&self, key: &ContextKey) -> Result<Arc<dyn Any>, ContextValueError> {
+        self.tree_node
+            .lock()
+            .unwrap()
+            .parent
+            .as_ref()
+            .expect("WithCancelは必ず親を持つ")
+            .upgrade()
+            .expect("親がすでに開放されています")
```

Comment:
これはどういうケースで発生しますか？

Reply:
基本的には発生しないとは思います。

`a -> b`のような親子関係があったとして、
bが変数に束縛されていたとして、aがdropした場合に起きます。

Reply:
ContextTreeの使われ方的に発生しないことが想定されるなら、expectの中の文字列は

```suggestion
            .expect("親から先に開放されることはない想定")
```

のように書いていたほうが、読み手にとってやさしいです。